atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java

```diff
@@ -243,11 +245,18 @@ private void schedulePollRequests() {
     }
 
     if (raft.getFirstCommitIndex() == 0 || raft.getState() == RaftContext.State.READY) {
-      final long missTime = System.currentTimeMillis() - lastHeartbeat;
+      final var leader =
+          Optional.ofNullable(raft.getLeader())
+              .map(DefaultRaftMember::memberId)
+              .map(MemberId::id)
+              .orElse("a known leader");
+
       log.info(
-          "No heartbeat from {} in the last {}ms, sending poll requests",
-          raft.getLeader(),
-          missTime);
+          "No heartbeat from {} after {} (time since last heartbeat in millis: {})",
+          leader,
+          electionTimer.getElectionTimeout(),
```

Comment:
My reasoning was they may not always be the same, and both are interesting information. When diagnosing issues for a cluster we don't operate, it can be useful to see the timeout being used (could be useful to spot bugs, for example), as well as the last time the heartbeat was received (which can be older than the last election time out, but if newer could indicate some bug in our scheduling). It can also inform us of issues, e.g. overloaded thread context queue, if the last heartbeat was a while ago but it took us quite some time to actually do the election. Since the timeout is typically randomized and we don't always know, it's hard to tell if there are major discrepancies.

That said, in the common case, they're probably the same, so I wouldn't insist here to keep it. If you feel it's just noise, we can remove it.

Reply:
In that case, I would rephrase it. 
` "No heartbeat from {} for that last {currentTime - lastHeartbeatTime} ms, election timeout is {}ms)"`

We can also initialize `lastHearbeatTime` to` System.currentTimeMillis()` in the constructor, so we don't need the check.

Also this only logs the configured election timeout, which is not really useful. We can see it from the startup logs. If you want to see the actual randomized delay, then electionTimer.getElectionTimeout() should return the current delay instead of the configured `electionTimeout`.

Reply:
Initializing it is definitely better, can't believe I didn't think of that :sweat_smile: I don't think it's particularly useful to distinguish if we have ever received one or not, so let's do that. And as for the election timeout, then my mistake, I thought the reset actually changed the election time out and that's what was used for the delay, but I admit I mostly just glanced at it and assumed it worked like this :sweat_smile: